rdrand_asm.S

```diff
@@ -23,6 +23,14 @@
 	.balign	64	; \
 	.globl	x	; \
 x:
+/* Add Intel CET IBT instrumentation */
+#if defined __CET__ && (__CET__ & 1)
+#ifdef __x86_64__
+	endbr64
+#elif defined(__i386__)
+	endbr32
+#endif
+#endif
```

Comment:
I see this code in all the examples I can find for endbr[32|64], but Im having trouble understanding why this endbr is needed in this asm code.  My understanding is that this needs to be the next instruction run at the target of any near/far call, and the target of any indirect jmp.  But looking at the asm code I don't see any of those in the code, so I'm unsure as to why the compiler is complaining about this code.

Edit: Wait, it might be for the call in the INIT_PIC macro

Reply:
A compiler is not complaining, it is an object verifier which reports IBT is not enabled in rdrand_asm.S and thus in the resulting rngd binary. I'm adding a section below (see below) to tell a verifier that rdrand_asm.S code supports IBT. But except _stating_ that IBT is supported we should actually _implement_ this support in code by adding ENDBR.

Yes, ENDBR should be the next instruction run after an _indirect_ near/far call and the target of any _indirect_ jmp. All the calls and jumps in this code are indeed direct. Calls to x86_rdrand_bytes() and others from C code are direct too as of now (I guess).

But in some future C code can be reworked to make an indirect call to these assembly functions. Or a compiler may decide to make a call from C code to x86_rdrand_bytes()/etc an indirect call for some reason (yes, this is unlikely). In this case and if IBT is active (it should be supported by OS and CPU and code) rngd process will be aborted without ENDBR being the first instruction of these assembly functions. So I'm adding ENDBR to the function prologue.

What puzzles me is the call to __x86.get_pc_thunk.bx. It is called directly and is local. But it is defined in a section marked both *.globl* and *.hidden*. I'm not sure what does this mean. Per my understanding .hidden means __x86.get_pc_thunk.bx cannot be seen and called outside of rdrand_asm.o object directly or indirectly. So we do not need ENDBR to __x86.get_pc_thunk.bx, as it is called only directly in rdrand_asm.S. I can be wrong here, so I'm asking for a review.

Reply:
yeah, this code was from before I took over the project.  From everything I can tell, this all looks reasonable.

Any thoughts on the 1f - 0f stuff below?